### PR TITLE
KAFKA-12932: Interfaces for SnapshotReader and SnapshotWriter

### PR DESCRIPTION
**Pull request by socutes**
Change the snapshot API so that SnapshotWriter and SnapshotReader are interfaces. Change the existing types SnapshotWriter and SnapshotReader to use a different name and to implement the interfaces introduced by this issue.

### CODE REVIEW

**Comment by socutes** [comment]
@showuon @jsancio please help review this pr. Thanks!

**Comment by socutes** [comment]
@dengziming

File: raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java [hunk 1]
```diff
@@ -17,218 +17,67 @@
...
-        } finally {
-            batches.forEach(CompletedBatch::release);
-        }
-    }
+    /**
+     * Adds a {@link SnapshotHeaderRecord} to snapshot
+     *
+     * @throws IllegalStateException if the snapshot is not empty
+     */
+    void initializeSnapshotWithHeader();
```
Comment (showuon) [inline raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java]: This method was originally `private` method. Why should we expose it as an interface method?

**Review by showuon — COMMENTED** [review]
@socutes , thanks for the PR, Left a comment. Thanks.

File: raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java (hunk 1 above)
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java]: Thanks for the warning. It really shouldn't be done. fixed! @showuon

**Comment by dengziming** [comment]
Currently we add a `FileSnapshotReader` class and copy some original code to it, a better way is to rename `SnapshotReader` to `FileSnapshotReader` and add a `SnapshotReader` Interface which can reduce code line changes tremendously.

**Comment by socutes** [comment]
Thank you for your advice. I know what you mean. At the code level, the final result should be the same. It's just that changing the number of lines of code might make it less, and I don't know if I have to.  @dengziming

**Pull request closed without merging**